onnxruntime/core/graph/graph.cc

```diff
@@ -2046,25 +2048,42 @@ void Graph::AddInitializedTensor(const TensorProto& tensor) {
 
     ORT_IGNORE_RETURN_VALUE(GetOrCreateNodeArg(tensor.name(), &t));
   }
-
-  SetGraphProtoSyncNeeded();
-  SetGraphResolveNeeded();
 }
 
 void Graph::RemoveInitializedTensor(const std::string& tensor_name) {
+  bool found = false;
   auto iter = name_to_initial_tensor_.find(tensor_name);
-  if (name_to_initial_tensor_.end() != iter) {
+  found = iter != name_to_initial_tensor_.end();
+  if (found) {
     name_to_initial_tensor_.erase(tensor_name);
-    SetGraphProtoSyncNeeded();
     SetGraphResolveNeeded();
   }
+
+  auto& mutable_initializers = *(graph_proto_->mutable_initializer());
+  auto proto_entry = std::find_if(mutable_initializers.begin(), mutable_initializers.end(),
+                                  [&tensor_name](const TensorProto& entry) { return entry.name() == tensor_name; });
+
+  if (proto_entry != mutable_initializers.end()) {
+    auto num_entries = mutable_initializers.size();
+    if (num_entries > 1) {
+      // move the last value into the one being removed, and delete the final entry
+      auto slot = proto_entry - mutable_initializers.begin();
+      auto last_entry = mutable_initializers.end() - 1;
+      mutable_initializers[slot] = std::move(*last_entry);
+      mutable_initializers.erase(last_entry);
+    } else {
+      mutable_initializers.erase(proto_entry);
+    }
+  } else {
+    // these should always be in sync as the pointer in name_to_initial_tensor_ is to memory in graph_proto_
```

Comment:
nit-picking about wording - would it be more accurate to say "the pointer in name_to_initial_tensor_ is to memory **owned by** graph_proto_"?